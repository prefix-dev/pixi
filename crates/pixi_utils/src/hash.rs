use rattler_digest::{Md5, Sha256, parse_digest_from_hex};
use rattler_lock::PackageHashes;
use std::fmt::Display;

use crate::errors::HashError;

/// Validates and parses a hash string into PackageHashes
///
/// # Arguments
/// * `algorithm` - The hash algorithm (e.g., "sha256", "md5")
/// * `hash_str` - The hash value as a hex string
/// * `package_name` - The package name for error messages
///
/// # Returns
/// * `Ok(PackageHashes)` if the hash is valid
/// * `Err(HashError)` if invalid
pub fn validate_and_parse_hash(
    algorithm: &str,
    hash_str: &str,
    package_name: &impl Display,
) -> Result<PackageHashes, HashError> {
    if hash_str.is_empty() {
        return Err(HashError::EmptyHash {
            algorithm: algorithm.to_uppercase(),
            package_name: package_name.to_string(),
        });
    }

    if !hash_str.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(HashError::InvalidHex {
            algorithm: algorithm.to_uppercase(),
            package_name: package_name.to_string(),
        });
    }

    match algorithm {
        "sha256" => {
            if hash_str.len() != 64 {
                return Err(HashError::InvalidLength {
                    algorithm: algorithm.to_uppercase(),
                    package_name: package_name.to_string(),
                    expected: 64,
                    actual: hash_str.len(),
                });
            }
            parse_digest_from_hex::<Sha256>(hash_str)
                .map(PackageHashes::Sha256)
                .ok_or_else(|| HashError::ParseFailed {
                    algorithm: algorithm.to_uppercase(),
                    package_name: package_name.to_string(),
                    hash_str: hash_str.to_string(),
                })
        }
        "md5" => {
            if hash_str.len() != 32 {
                return Err(HashError::InvalidLength {
                    algorithm: algorithm.to_uppercase(),
                    package_name: package_name.to_string(),
                    expected: 32,
                    actual: hash_str.len(),
                });
            }
            parse_digest_from_hex::<Md5>(hash_str)
                .map(PackageHashes::Md5)
                .ok_or_else(|| HashError::ParseFailed {
                    algorithm: algorithm.to_uppercase(),
                    package_name: package_name.to_string(),
                    hash_str: hash_str.to_string(),
                })
        }
        _ => Err(HashError::UnsupportedAlgorithm {
            algorithm: algorithm.to_string(),
            package_name: package_name.to_string(),
        }),
    }
}

/// Extracts and parses a hash from a URL fragment
///
/// Parses fragments like "#sha256=abc123" or "#md5=def456&egg=foo"
///
/// # Arguments
/// * `fragment` - The URL fragment (without the leading #)
/// * `package_name` - The package name for error messages
///
/// # Returns
/// * `Ok(Some(PackageHashes))` if a valid hash is found
/// * `Ok(None)` if no hash parameter is found
/// * `Err(HashError)` if a hash parameter is found but invalid
pub fn parse_hash_from_url_fragment(
    fragment: &str,
    package_name: &impl Display,
) -> Result<Option<PackageHashes>, HashError> {
    for param in fragment.split('&') {
        if let Some((algorithm, hash_str)) = param.split_once('=') {
            let algorithm_lower = algorithm.to_lowercase();

            match algorithm_lower.as_str() {
                "sha256" | "md5" => {
                    return validate_and_parse_hash(&algorithm_lower, hash_str, package_name)
                        .map(Some);
                }
                alg if alg.contains("sha") || alg.contains("md5") || alg.contains("blake") => {
                    return Err(HashError::UnsupportedAlgorithm {
                        algorithm: algorithm_lower,
                        package_name: package_name.to_string(),
                    });
                }
                _ => continue,
            }
        }
    }

    Ok(None)
}

/// Updates or adds a hash parameter to a URL fragment while preserving other parameters
///
/// For example:
/// - "egg=foo" + sha256 hash -> "egg=foo&sha256=abc123"
/// - "sha256=old&egg=foo" + new sha256 -> "sha256=new&egg=foo"
/// - "" + sha256 hash -> "sha256=abc123"
pub fn update_fragment_with_hash(fragment: Option<&str>, hash: &PackageHashes) -> String {
    // Convert hash to fragment format
    let hash_param = match hash {
        PackageHashes::Sha256(sha256) => format!("sha256={:x}", sha256),
        PackageHashes::Md5(md5) => format!("md5={:x}", md5),
        PackageHashes::Md5Sha256(_md5, sha256) => {
            // Prefer SHA256 for direct URLs
            format!("sha256={:x}", sha256)
        }
    };

    match fragment {
        None | Some("") => hash_param,
        Some(existing) => {
            // Parse existing parameters
            let mut params: Vec<String> = existing.split('&').map(|s| s.to_string()).collect();
            let mut hash_updated = false;

            // Update existing hash parameter if present
            for param in &mut params {
                if param.starts_with("sha256=") || param.starts_with("md5=") {
                    *param = hash_param.clone();
                    hash_updated = true;
                    break;
                }
            }

            // If no hash parameter was found, add it
            if !hash_updated {
                params.push(hash_param);
            }

            params.join("&")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_hash_from_url_fragment() {
        // Valid SHA256
        let result = parse_hash_from_url_fragment(
            "sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
            &"test-package",
        )
        .unwrap();
        assert!(matches!(result, Some(PackageHashes::Sha256(_))));

        // Valid MD5
        let result =
            parse_hash_from_url_fragment("md5=d41d8cd98f00b204e9800998ecf8427e", &"test-package")
                .unwrap();
        assert!(matches!(result, Some(PackageHashes::Md5(_))));

        // Multiple parameters with hash
        let result = parse_hash_from_url_fragment(
            "egg=foo&sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855&bar=baz",
            &"test-package"
        ).unwrap();
        assert!(matches!(result, Some(PackageHashes::Sha256(_))));

        // No hash parameter
        let result = parse_hash_from_url_fragment("egg=foo&bar=baz", &"test-package").unwrap();
        assert!(result.is_none());

        // Invalid hash algorithm
        let result = parse_hash_from_url_fragment("sha512=abcdef", &"test-package");
        assert!(result.is_err());

        // Empty hash
        let result = parse_hash_from_url_fragment("sha256=", &"test-package");
        assert!(result.is_err());

        // Invalid hex
        let result = parse_hash_from_url_fragment("sha256=xyz", &"test-package");
        assert!(result.is_err());
    }

    #[test]
    fn test_update_fragment_with_hash() {
        let sha256_hash = parse_digest_from_hex::<Sha256>(
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        )
        .unwrap();
        let md5_hash = parse_digest_from_hex::<Md5>("d41d8cd98f00b204e9800998ecf8427e").unwrap();

        // Test 1: No existing fragment
        let result = update_fragment_with_hash(None, &PackageHashes::Sha256(sha256_hash));
        assert_eq!(
            result,
            "sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );

        // Test 2: Empty fragment
        let result = update_fragment_with_hash(Some(""), &PackageHashes::Sha256(sha256_hash));
        assert_eq!(
            result,
            "sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );

        // Test 3: Fragment with egg parameter
        let result =
            update_fragment_with_hash(Some("egg=mypackage"), &PackageHashes::Sha256(sha256_hash));
        assert_eq!(
            result,
            "egg=mypackage&sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );

        // Test 4: Fragment with multiple parameters
        let result = update_fragment_with_hash(
            Some("egg=mypackage&subdirectory=src"),
            &PackageHashes::Sha256(sha256_hash),
        );
        assert_eq!(
            result,
            "egg=mypackage&subdirectory=src&sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );

        // Test 5: Fragment with existing sha256 hash
        let result = update_fragment_with_hash(
            Some("sha256=oldhash&egg=mypackage"),
            &PackageHashes::Sha256(sha256_hash),
        );
        assert_eq!(
            result,
            "sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855&egg=mypackage"
        );

        // Test 6: Fragment with existing md5 hash, updating with sha256
        let result = update_fragment_with_hash(
            Some("md5=oldhash&egg=mypackage"),
            &PackageHashes::Sha256(sha256_hash),
        );
        assert_eq!(
            result,
            "sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855&egg=mypackage"
        );

        // Test 7: MD5 hash
        let result =
            update_fragment_with_hash(Some("egg=mypackage"), &PackageHashes::Md5(md5_hash));
        assert_eq!(result, "egg=mypackage&md5=d41d8cd98f00b204e9800998ecf8427e");

        // Test 8: Md5Sha256 hash (should prefer SHA256)
        let result = update_fragment_with_hash(
            Some("egg=mypackage"),
            &PackageHashes::Md5Sha256(md5_hash, sha256_hash),
        );
        assert_eq!(
            result,
            "egg=mypackage&sha256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );
    }
}
