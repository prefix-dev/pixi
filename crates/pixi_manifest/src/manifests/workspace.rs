use std::hash::Hash;

use indexmap::{map::IndexMap, Equivalent};

use crate::{
    consts,
    environment::{Environment, EnvironmentName},
    environments::Environments,
    feature::{Feature, FeatureName},
    solve_group::SolveGroups,
    toml::ExternalWorkspaceProperties,
    workspace::Workspace,
    TomlError,
};

/// Holds the parsed content of the workspace part of a pixi manifest. This
/// describes the part related to the workspace only.
#[derive(Debug, Clone)]
pub struct WorkspaceManifest {
    /// Information about the project
    pub workspace: Workspace,

    /// All the features defined in the project.
    pub features: IndexMap<FeatureName, Feature>,

    /// All the environments defined in the project.
    pub environments: Environments,

    /// The solve groups that are part of the project.
    pub solve_groups: SolveGroups,
}

impl WorkspaceManifest {
    /// Parses a TOML string into a `WorkspaceManifest`.
    pub fn from_toml_str(toml_str: &str) -> Result<Self, TomlError> {
        let manifest = crate::toml::TomlManifest::from_toml_str(toml_str)?;
        Ok(manifest
            .into_manifests(ExternalWorkspaceProperties::default())?
            .0)
    }

    /// Returns the default feature.
    ///
    /// This is the feature that is added implicitly by the tables at the root
    /// of the project manifest.
    pub fn default_feature(&self) -> &Feature {
        self.features
            .get(&FeatureName::Default)
            .expect("default feature should always exist")
    }

    /// Returns a mutable reference to the default feature.
    pub(crate) fn default_feature_mut(&mut self) -> &mut Feature {
        self.features
            .get_mut(&FeatureName::Default)
            .expect("default feature should always exist")
    }

    /// Returns the default environment
    ///
    /// This is the environment that is added implicitly as the environment with
    /// only the default feature. The default environment can be overwritten
    /// by a environment named `default`.
    pub fn default_environment(&self) -> &Environment {
        let envs = &self.environments;
        envs.find(&EnvironmentName::Named(String::from(
            consts::DEFAULT_ENVIRONMENT_NAME,
        )))
        .or_else(|| envs.find(&EnvironmentName::Default))
        .expect("default environment should always exist")
    }

    /// Returns the environment with the given name or `None` if it does not
    /// exist.
    pub fn environment<Q>(&self, name: &Q) -> Option<&Environment>
    where
        Q: ?Sized + Hash + Equivalent<EnvironmentName>,
    {
        self.environments.find(name)
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot, assert_yaml_snapshot};
    use itertools::Itertools;
    use rattler_conda_types::{NamedChannelOrUrl, Platform};

    use crate::{utils::test_utils::expect_parse_failure, TargetSelector, WorkspaceManifest};

    const PROJECT_BOILERPLATE: &str = r#"
        [project]
        name = "foo"
        version = "0.1.0"
        channels = []
        platforms = []
        "#;

    #[test]
    fn test_target_specific() {
        let contents = format!(
            r#"
        {PROJECT_BOILERPLATE}

        [target.win-64.dependencies]
        foo = "3.4.5"

        [target.osx-64.dependencies]
        foo = "1.2.3"
        "#
        );

        let manifest = WorkspaceManifest::from_toml_str(&contents).unwrap();
        let targets = &manifest.default_feature().targets;
        assert_eq!(
            targets.user_defined_selectors().cloned().collect_vec(),
            vec![
                TargetSelector::Platform(Platform::Win64),
                TargetSelector::Platform(Platform::Osx64),
            ]
        );

        let win64_target = targets
            .for_target(&TargetSelector::Platform(Platform::Win64))
            .unwrap();
        let osx64_target = targets
            .for_target(&TargetSelector::Platform(Platform::Osx64))
            .unwrap();
        assert_eq!(
            win64_target
                .run_dependencies()
                .unwrap()
                .get("foo")
                .unwrap()
                .as_version_spec()
                .unwrap()
                .to_string(),
            "==3.4.5"
        );
        assert_eq!(
            osx64_target
                .run_dependencies()
                .unwrap()
                .get("foo")
                .unwrap()
                .as_version_spec()
                .unwrap()
                .to_string(),
            "==1.2.3"
        );
    }

    #[test]
    fn test_mapped_dependencies() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [dependencies]
            test_map = {{ version = ">=1.2.3", channel="conda-forge", build="py34_0" }}
            test_build = {{ build = "bla" }}
            test_channel = {{ channel = "conda-forge" }}
            test_version = {{ version = ">=1.2.3" }}
            test_version_channel = {{ version = ">=1.2.3", channel = "conda-forge" }}
            test_version_build = {{ version = ">=1.2.3", build = "py34_0" }}
            "#
        );

        let manifest = WorkspaceManifest::from_toml_str(&contents).unwrap();
        let deps = manifest
            .default_feature()
            .targets
            .default()
            .run_dependencies()
            .unwrap();
        let test_map_spec = deps.get("test_map").unwrap().as_detailed().unwrap();

        assert_eq!(
            test_map_spec.version.as_ref().unwrap().to_string(),
            ">=1.2.3"
        );
        assert_eq!(test_map_spec.build.as_ref().unwrap().to_string(), "py34_0");
        assert_eq!(
            test_map_spec.channel,
            Some(NamedChannelOrUrl::Name("conda-forge".to_string()))
        );

        assert_eq!(
            deps.get("test_build")
                .unwrap()
                .as_detailed()
                .unwrap()
                .build
                .as_ref()
                .unwrap()
                .to_string(),
            "bla"
        );

        let test_channel = deps.get("test_channel").unwrap().as_detailed().unwrap();
        assert_eq!(
            test_channel.channel,
            Some(NamedChannelOrUrl::Name("conda-forge".to_string()))
        );

        let test_version = deps.get("test_version").unwrap().as_detailed().unwrap();
        assert_eq!(
            test_version.version.as_ref().unwrap().to_string(),
            ">=1.2.3"
        );

        let test_version_channel = deps
            .get("test_version_channel")
            .unwrap()
            .as_detailed()
            .unwrap();
        assert_eq!(
            test_version_channel.version.as_ref().unwrap().to_string(),
            ">=1.2.3"
        );
        assert_eq!(
            test_version_channel.channel,
            Some(NamedChannelOrUrl::Name("conda-forge".to_string()))
        );

        let test_version_build = deps
            .get("test_version_build")
            .unwrap()
            .as_detailed()
            .unwrap();
        assert_eq!(
            test_version_build.version.as_ref().unwrap().to_string(),
            ">=1.2.3"
        );
        assert_eq!(
            test_version_build.build.as_ref().unwrap().to_string(),
            "py34_0"
        );
    }

    #[test]
    fn test_dependency_types() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [dependencies]
            my-game = "1.0.0"

            [build-dependencies]
            cmake = "*"

            [host-dependencies]
            sdl2 = "*"
            "#
        );

        let manifest = WorkspaceManifest::from_toml_str(&contents).unwrap();
        let default_target = manifest.default_feature().targets.default();
        let run_dependencies = default_target.run_dependencies().unwrap();
        let build_dependencies = default_target.build_dependencies().unwrap();
        let host_dependencies = default_target.host_dependencies().unwrap();

        assert_eq!(
            run_dependencies
                .get("my-game")
                .unwrap()
                .as_version_spec()
                .unwrap()
                .to_string(),
            "==1.0.0"
        );
        assert_eq!(
            build_dependencies
                .get("cmake")
                .unwrap()
                .as_version_spec()
                .unwrap()
                .to_string(),
            "*"
        );
        assert_eq!(
            host_dependencies
                .get("sdl2")
                .unwrap()
                .as_version_spec()
                .unwrap()
                .to_string(),
            "*"
        );
    }

    #[test]
    fn test_invalid_target_specific() {
        let examples = [r#"[target.foobar.dependencies]
            invalid_platform = "henk""#];

        assert_snapshot!(expect_parse_failure(&format!(
            "{PROJECT_BOILERPLATE}\n{}",
            examples[0]
        )));
    }

    #[test]
    fn test_invalid_key() {
        insta::with_settings!({snapshot_suffix => "foobar"}, {
            assert_snapshot!(expect_parse_failure(&format!("{PROJECT_BOILERPLATE}\n[foobar]")))
        });

        insta::with_settings!({snapshot_suffix => "hostdependencies"}, {
            assert_snapshot!(expect_parse_failure(&format!("{PROJECT_BOILERPLATE}\n[target.win-64.hostdependencies]")))
        });

        insta::with_settings!({snapshot_suffix => "environment"}, {
            assert_snapshot!(expect_parse_failure(&format!("{PROJECT_BOILERPLATE}\n[environments.INVALID]")))
        });
    }

    #[test]
    fn test_target_specific_tasks() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [tasks]
            test = "test multi"

            [target.win-64.tasks]
            test = "test win"

            [target.linux-64.tasks]
            test = "test linux"
            "#
        );

        let manifest = WorkspaceManifest::from_toml_str(&contents).unwrap();

        assert_snapshot!(manifest
            .default_feature()
            .targets
            .iter()
            .flat_map(|(target, selector)| {
                let selector_name =
                    selector.map_or_else(|| String::from("default"), ToString::to_string);
                target.tasks.iter().filter_map(move |(name, task)| {
                    Some(format!(
                        "{}/{} = {}",
                        &selector_name,
                        name.as_str(),
                        task.as_single_command()?
                    ))
                })
            })
            .join("\n"));
    }

    #[test]
    fn test_python_dependencies() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [pypi-dependencies]
            foo = ">=3.12"
            bar = {{ version=">=3.12", extras=["baz"] }}
            "#
        );

        assert_snapshot!(WorkspaceManifest::from_toml_str(&contents)
            .expect("parsing should succeed!")
            .default_feature()
            .targets
            .default()
            .pypi_dependencies
            .clone()
            .into_iter()
            .flat_map(|d| d.into_iter())
            .map(|(name, spec)| format!("{} = {}", name.as_source(), toml_edit::Value::from(spec)))
            .join("\n"));
    }

    #[test]
    fn test_pypi_options_default_feature() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [project.pypi-options]
            index-url = "https://pypi.org/simple"
            extra-index-urls = ["https://pypi.org/simple2"]
            [[project.pypi-options.find-links]]
            path = "../foo"
            [[project.pypi-options.find-links]]
            url = "https://example.com/bar"
            "#
        );

        assert_yaml_snapshot!(WorkspaceManifest::from_toml_str(&contents)
            .expect("parsing should succeed!")
            .workspace
            .pypi_options
            .clone()
            .unwrap());
    }

    #[test]
    fn test_pypy_options_project_and_default_feature() {
        let contents = format!(
            r#"
            {PROJECT_BOILERPLATE}
            [project.pypi-options]
            extra-index-urls = ["https://pypi.org/simple2"]

            [pypi-options]
            extra-index-urls = ["https://pypi.org/simple3"]
            "#
        );

        let manifest =
            WorkspaceManifest::from_toml_str(&contents).expect("parsing should succeed!");
        assert_yaml_snapshot!(manifest.workspace.pypi_options.clone().unwrap());
    }

    #[test]
    fn test_duplicate_dependency() {
        let contents = format!(
            r#"
        {PROJECT_BOILERPLATE}

        [dependencies]
        Flask = "2.*"
        flask = "2.*"
        "#
        );
        let manifest = WorkspaceManifest::from_toml_str(&contents);

        assert!(manifest.is_err());
        assert!(manifest
            .unwrap_err()
            .to_string()
            .contains("duplicate dependency"));
    }

    #[test]
    fn test_duplicate_host_dependency() {
        let contents = format!(
            r#"
        {PROJECT_BOILERPLATE}

        [host-dependencies]
        LibC = "2.12"
        libc = "2.12"
        "#
        );
        let manifest = WorkspaceManifest::from_toml_str(&contents);

        assert!(manifest.is_err());
        assert!(manifest
            .unwrap_err()
            .to_string()
            .contains("duplicate dependency"));
    }

    #[test]
    fn test_tool_deserialization() {
        let contents = r#"
        [project]
        name = "foo"
        channels = []
        platforms = []
        [tool.ruff]
        test = "test"
        test1 = ["test"]
        test2 = { test = "test" }

        [tool.ruff.test3]
        test = "test"

        [tool.poetry]
        test = "test"
        "#;
        let _manifest = WorkspaceManifest::from_toml_str(contents).unwrap();
    }
}
