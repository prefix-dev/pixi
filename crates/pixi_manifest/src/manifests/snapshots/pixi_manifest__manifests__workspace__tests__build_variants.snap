---
source: crates/pixi_manifest/src/manifests/workspace.rs
expression: resolved_linux
---
[
    Some(
        {
            "python": [
                "3.10.*",
                "3.11.*",
            ],
        },
    ),
]
